src/include/homestore/blk.h

```diff
@@ -24,122 +24,162 @@
 
-public:
-    static constexpr blk_count_t max_blks_in_op() { return (1 << NBLKS_BITS); }
-    static constexpr uint64_t max_id_int() { return (1ull << (BLK_NUM_BITS + NBLKS_BITS + CHUNK_NUM_BITS)) - 1; }
+        serialized() : m_is_multi{0}, m_blk_num{0}, m_nblks{0}, m_chunk_num{0} {}
+        serialized(bool is_multi, blk_num_t blk_num, blk_count_t nblks, chunk_num_t cnum) :
+                m_is_multi{is_multi ? 0x1u : 0x0u}, m_blk_num{blk_num}, m_nblks{nblks}, m_chunk_num{cnum} {}
+    };
+    static_assert(sizeof(serialized) == sizeof(uint64_t), "Expected serialized size to 64 bits");
 
-    static int compare(const BlkId& one, const BlkId& two);
-    uint64_t to_integer() const;
+    serialized s;
 
+public:
+    BlkId() = default;
     explicit BlkId(uint64_t id_int);
-    BlkId(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num = 0);
-    BlkId() { invalidate(); }
-    BlkId(const BlkId&) = default;
-    BlkId& operator=(const BlkId&) = default;
+    BlkId(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num);
+    BlkId(BlkId const&) = default;
+    BlkId& operator=(BlkId const&) = default;
     BlkId(BlkId&&) noexcept = default;
     BlkId& operator=(BlkId&&) noexcept = default;
-    bool operator==(const BlkId& other) const { return (compare(*this, other) == 0); }
-    bool operator>(const BlkId& other) const { return (compare(*this, other) > 0); }
-    bool operator<(const BlkId& other) const { return (compare(*this, other) < 0); }
+
+    bool operator==(BlkId const& other) const { return (compare(*this, other) == 0); }
+    bool operator>(BlkId const& other) const { return (compare(*this, other) > 0); }
+    bool operator<(BlkId const& other) const { return (compare(*this, other) < 0); }
+
+    blk_num_t blk_num() const { return s.m_blk_num; }
+    blk_count_t blk_count() const { return s.m_nblks; }
+    chunk_num_t chunk_num() const { return s.m_chunk_num; }
+    bool is_multi() const { return s.m_is_multi; }
 
     void invalidate();
+    uint64_t to_integer() const;
+    sisl::blob serialize();
+    void deserialize(sisl::blob const& b, bool copy);
+    uint32_t serialized_size() const;
+    std::string to_string() const;
     bool is_valid() const;
 
-    BlkId get_blkid_at(uint32_t offset, uint32_t pagesz) const;
-    BlkId get_blkid_at(uint32_t offset, uint32_t size, uint32_t pagesz) const;
-
-    void set(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num = 0);
-    void set(const BlkId& bid);
-    void set(uint64_t id_int);
+    static int compare(BlkId const& one, BlkId const& two);
+};
+#pragma pack()
 
-    void set_blk_num(blk_num_t blk_num);
-    blk_num_t get_blk_num() const { return m_blk_num; }
-    // last blk num is the last blk num that belongs this blkid;
-    blk_num_t get_last_blk_num() const { return get_blk_num() + get_nblks() - 1; }
+#pragma pack(1)
+struct MultiBlkId : public BlkId {
+    static constexpr uint32_t max_addln_pieces{5};
```

Comment:
Is there a reason for max_addln_pieces to be such a small number?
Or is it because we don't want to dynamic extending of the pieces?

Reply:
Having higher number will increase the memory foot print, because it is freely passed around. As is they are 48 bytes. If a block is not available in 6 pieces, then its better to return the allocation failure than doing it otherwise IMO.